py/obj.h

```diff
@@ -616,15 +632,18 @@ struct _mp_obj_type_t {
     // Can return MP_OBJ_NULL if operation not supported.
     uint8_t slot_index_subscr;
 
-    // Corresponds to __iter__ special method.
-    // Can use the given mp_obj_iter_buf_t to store iterator object,
-    // otherwise can return a pointer to an object on the heap.
+    // This slot's behavior depends on the MP_TYPE_FLAG_ITERNEXT-* flags above.
+    // - If no MP_TYPE_FLAG_ITERNEXT-* flags are set, then this corresponds to __iter__
+    //   special method (of type mp_getiter_fun_t). Can use the given mp_obj_iter_buf_t
+    //   to store the iterator object, otherwise can return a pointer to an object on the heap.
+    // - If MP_TYPE_FLAG_ITERNEXT is set, then this corresponds to __next__ special method.
+    //   May return MP_OBJ_STOP_ITERATION as an optimisation instead of raising StopIteration()
+    //   with no args. The type will implicitly implement getiter as "return self".
+    // - If MP_TYPE_FLAG_ITERNEXT_CUSTOM is set, then this slot must point to an
+    //   mp_getiter_iternext_custom_t instance with both the getiter and iternext fields set.
+    // - If MP_TYPE_FLAG_ITERNEXT_STREAM is set, this this slot should be unset.
     uint8_t slot_index_getiter;
```

Comment:
I feel like this slot should be renamed to `slot_index_iter`, to indicate that it's generally used for iteration.  Then could have the following flags:
```c
#define MP_TYPE_FLAG_ITER_IS_GETITER (0x0000)
#define MP_TYPE_FLAG_ITER_IS_ITERNEXT (0x0080)
#define MP_TYPE_FLAG_ITER_IS_CUSTOM (0x0100)
#define MP_TYPE_FLAG_ITER_IS_STREAM (...)
```
?

Reply:
Yep good idea. Updated.